src/libproxychains.c

```diff
@@ -587,6 +600,51 @@ static int is_v4inv6(const struct in6_addr *a) {
 	return !memcmp(a->s6_addr, "\0\0\0\0\0\0\0\0\0\0\xff\xff", 12);
 }
 
+static int compare_func_int(const void *l, const void *r) {
+    int arg1 = *(const int*)l;
+    int arg2 = *(const int*)r;
+    return (arg1 > arg2) - (arg1 < arg2);
+}
+/* Warning: Linux manual says the third arg is `unsigned int`, but unistd.h says `int`. */
+HOOKFUNC(int, close_range, unsigned first, unsigned last, int flags) {
+    if(!init_l) {
+        /* push back to cache, and delay the execution. */
+        if(close_range_buffer_cnt >= (sizeof close_range_buffer / sizeof *close_range_buffer)) {
+            errno = ENOMEM;
+            return -1;
+        }
+        int i = close_range_buffer_cnt++;
+        close_range_buffer[i].first = first;
+        close_range_buffer[i].last = last;
+        close_range_buffer[i].flags = flags;
+        return errno = 0;
+    }
+	if(proxychains_resolver != DNSLF_RDNS_THREAD) return true_close_range(first, last, flags);
+
+	/* prevent rude programs (like ssh) from closing our pipes */
+    int res = 0, uerrno = 0;
+    int protected_fds[] = {req_pipefd[0], req_pipefd[1], resp_pipefd[0], resp_pipefd[1]};
+    int protected_fds_size = sizeof protected_fds / sizeof *protected_fds;
+    qsort(protected_fds, protected_fds_size, sizeof *protected_fds, compare_func_int);
+    /* We are skipping protected_fds while calling true_close_range() */
+    for(int i = 0; i < protected_fds_size; ++i) {
+        if(protected_fds[i] < first || protected_fds[i] > last)
+            continue;
+        int prev = (i == 0 || protected_fds[i-1] < first) ? first : protected_fds[i-1]+1;
+        int next = (i == protected_fds_size-1 || protected_fds[i+1] > last) ? last : protected_fds[i+1]-1;
+        if(prev != protected_fds[i]) {
+		    res |= true_close_range(prev, protected_fds[i]-1, flags);
+            uerrno = errno == 0 ? uerrno : errno;
```

Comment:
you should only access errno iff the call to true_close_range returned error in that specific invocation

Reply:
fixed in latest commit